### PR TITLE
add c2st metric + tests

### PR DESCRIPTION
**Pull request by Justinezgh**

### CODE REVIEW

**Comment by EiffL** [comment]
Minor comment: Don't forget to add at least a few lines of description in your PRs ;-) so that the reviewer can know a bit what to expect.

File: sbids/metrics/c2st.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import numpy as np
+import jax
+import jax.numpy as jnp
+import haiku as hk
+import optax
+from tensorflow_probability.substrates import jax as tfp
+tfd = tfp.distributions
+tfb = tfp.bijectors
+
+class Classifier_C2ST(hk.Module):
```
Comment (EiffL) [inline sbids/metrics/c2st.py:12]: Is this following the architecture of the SBIBM paper? If so, we can add one line of comment here to say what we based our model on.

File: sbids/metrics/c2st.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    net_x = jax.nn.leaky_relu(net_x) 
+    net_x = hk.Linear(32)(net_x) 
+    net_x = jax.nn.leaky_relu(net_x) 
+    net_x = hk.Linear(16)(net_x) 
+    net_x = jax.nn.leaky_relu(net_x) 
+    net_x = hk.Linear(8)(net_x) 
+    net_x = jax.nn.leaky_relu(net_x) 
+    net_x = hk.Linear(1)(net_x) 
+    net_x = jax.nn.sigmoid(net_x).squeeze()
+    net_x = jnp.clip(net_x, 1e-5, 1.-1e-5)
```
Comment (EiffL) [inline sbids/metrics/c2st.py:28]: is this clipping line really needed?

File: sbids/metrics/c2st.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    return net_x
+
+
+def c2st(sample1,sample2,num_seed):
```
Comment (EiffL) [inline sbids/metrics/c2st.py:33]: same comment as before ^^, never pass a num_seed as argument to a jax function, but an actual prngkey

File: sbids/metrics/c2st.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  sample1_std = jnp.std(sample1, axis = 0)
+  sample1 = (sample1 - sample1_mean) / sample1_std
+  sample2 = (sample2 - sample1_mean) / sample1_std
+
+  label1 = jnp.zeros(len(sample1)) 
+  label2 = jnp.ones(len(sample2))  
+  data = jnp.concatenate([jnp.concatenate([sample1,label1.reshape(-1,1)],axis = 1),
+                          jnp.concatenate([sample2,label2.reshape(-1,1)],axis = 1)],axis = 0)
+
+  data = jax.random.permutation(jax.random.PRNGKey(num_seed), data, axis=0)
```
Comment (EiffL) [inline sbids/metrics/c2st.py:62]: use key from input

File: sbids/metrics/c2st.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  optimizer_c2stClassifier = optax.chain(optax.scale_by_adam(), optax.scale_by_schedule(scheduler), optax.scale(-1))
+  opt_state_c2stClassifier = optimizer_c2stClassifier.init(params_c2stClassifier)
+
+
+  @jax.jit
+  def logloss(params,batch,label):
+      """Loss"""
+      prob = jax.vmap(lambda x : c2stClassifier.apply(params,x.reshape(4)))(batch)
+      test = (1-label)*jnp.log(1-prob)
+      return -jnp.mean(label*jnp.log(prob)+(1-label)*jnp.log(1-prob))
```
Comment (EiffL) [inline sbids/metrics/c2st.py:102]: giphy

:-| this is not wrong, but never implement a loss function involving log and/or exp by hand, it can easily be unstable because of overflow/underflow issues. Always try to use an exisiting function from a NN library because it usually implements stability tricks.

In this particular case, you can use this one from optax https://optax.readthedocs.io/en/latest/api.html#optax.sigmoid_binary_cross_entropy

File: sbids/metrics/c2st.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  data = jnp.concatenate([jnp.concatenate([sample1,label1.reshape(-1,1)],axis = 1),
+                          jnp.concatenate([sample2,label2.reshape(-1,1)],axis = 1)],axis = 0)
+
+  data = jax.random.permutation(jax.random.PRNGKey(num_seed), data, axis=0)
+  ntrain = int((len(sample1)+len(sample2))*0.6)
+  dtrain, dte = data[:ntrain], data[ntrain:]
+
+  sample_dim = sample1.shape[1]
+  ztrain = dtrain[...,:sample_dim]
+  ltrain = dtrain[...,sample_dim:].squeeze()
```
Comment (EiffL) [inline sbids/metrics/c2st.py:68]: I don't understand why you concatenate the data first, and then split it again. What's the reason for that?

Comment (EiffL) [reply sbids/metrics/c2st.py:68]: you don't need to shuffle at the top. Just use 2 permutations in you `data_stream` function

**Review by EiffL — COMMENTED** [review]
Just a few comments, but looks good otherwise

File: sbids/metrics/c2st.py (hunk 1 above)
Comment (Justinezgh) [reply sbids/metrics/c2st.py:12]: nop it's not the same, their classifier: 
```python 
    clf = MLPClassifier(
        activation="relu",
        hidden_layer_sizes=(10 * ndim, 10 * ndim),
        max_iter=10000,
        solver="adam",
        random_state=seed,
    )
```

File: sbids/metrics/c2st.py (hunk 2 above)
Comment (Justinezgh) [reply sbids/metrics/c2st.py:28]: I think so otherwise I got probabilities of 0 or 1 and so nans in the loss because of the log()

File: sbids/metrics/c2st.py (hunk 1 above)
Comment (EiffL) [reply sbids/metrics/c2st.py:12]: So, why is it different ^^?

File: sbids/metrics/c2st.py (hunk 2 above)
Comment (EiffL) [reply sbids/metrics/c2st.py:28]: Not if you use the loss I mention below

Comment (Justinezgh) [reply sbids/metrics/c2st.py:28]: ok :)

File: sbids/metrics/c2st.py (hunk 6 above)
Comment (Justinezgh) [reply sbids/metrics/c2st.py:68]: not sure to understand, but it was to create data train and data test

File: sbids/metrics/c2st.py (hunk 1 above)
Comment (Justinezgh) [reply sbids/metrics/c2st.py:12]: seems that I have good results now, I change it

File: sbids/metrics/c2st.py (hunk 6 above)
Comment (EiffL) [reply sbids/metrics/c2st.py]: oh, I see, yeah because it's after shuffling.

Yeah so my point was more that I wasnt surre concatenating both samples and shuffling them was the best way to do this.  Since you shuffle again below. detail. dtail

Comment (EiffL) [reply sbids/metrics/c2st.py]: not very important as long as the 2 saamplees are exactly of the same size
